### ISSUE TITLE
Kryo breaks down while serializing genomic data

### ISSUE THREAD

**Issue by mulugetam** [EsotericSoftware/kryo#382]
**Bug**:
While running a genomic pipeline workload over a Spark cluster, we found out that Kryo breaks down throwing a `java.lang.NegativeArraySizeException` exception. The stack trace indicates that it happens when Spark attempts to blockify our genomic object using Kryo.

**Cause**:
The stash in the Cuckoo hash implementation of `IdentityObjectIntMap `fills up rapidly initiating a doubling of the hash table capacity beyond `Integer.MAX_VALUE`. 

We were able to isolate the problem object and serialize it out (using Java serialization) to a file of size 743M and reproduce the problem independently of our running context (We wrote a simple test application that de-serializes our object file using Java serialization and then attempt to serialize it using Kryo). 

The memory footprint of our genomic object (https://github.com/DimitrisAndreou/memory-measurer):

- size: 1499106912 bytes (1.4G)
- # objects: 57,652,339
- # primitives: int x 57688824, boolean x 57652108, char x 74
- # references: 86,478,445

Further debugging reveals that the stash size fills up for a load factor as low as 5%:

**Rudimentary work around**:
The rudimentary work around we took to temporarily solve the problem is to gradually increase the _stashCapacity_ based on the stash load factor.

_Current implementation_:

```
private void resize (int newSize) {
…
stashCapacity = Math.max(3, (int)Math.ceil(Math.log(capacity)) * 2); 
… 
}
```

_Changed to_:
 
```
private void resize (int newSize) {
…
stashCapacity = Math.max(3,(int)(Math.ceil(Math.log(capacity)) * (2.0 + stashSize/stashCapacity))); 
… 
}
```

In the worst and most extreme case the change would increase the `stashCapacity `from about 44 elements (current implementation) to about 65 elements. And this has temporarily fixed the bug for us.

Attached you may find the raw data from our runs using different load factors (all of which break down for the current implementation). Also, since Kryo is being used to blockify other objects inside Spark, it should NOT break for the default load factor (0.8) and initial capacity (32). 

[kryo_bug_raw.xlsx](https://github.com/EsotericSoftware/kryo/files/113379/kryo_bug_raw.xlsx)

Thanks,
Mulugeta

**Comment by magro** [comment]
@mulugetam Great analysis! Do you want to submit a pull request?

**Comment by magro** [comment]
@mulugetam Is this still an issue for you or shall we close this?

**Comment by mmammo** [comment]
@magro it is still an issue. The change I proposed in the stash formula fixed the issue for some of the genome data sets, but not for others. The stash simply fills up quickly due to too many collisions. Increasing stash size to as much as *8 (instead of the current 2) didn't help either. 

The best solution I found now is to throw in a fourth hash function if the capacity (i.e. the hash table size) grows very large (e.g. 2^15, from the experiments I did).  I have benchmarked all my experiments using [jmh](http://openjdk.java.net/projects/code-tools/jmh/) from openjdk (at microseconds level) and the performance (for both put and get) look promising. Also, I have tested it for both genome as well as randomly generated data sets. 

I will share you all my benchmark results over the weekend or early next week. I will also send you my proposed patch as soon as I complete an internal process here.

Thanks,
Mulugeta

**Comment by magro** [comment]
@mmammo Looking forward to your results and the pull request! Thanks!

**Comment by mikism** [comment]
Hi. We also have this problem. However, it manifests only when running code on JRE 8, with the same NegativeArraySizeException. But JRE 7 works fine (compiling with JDK 7 or 8 does not make any difference). My guess is that identity hash code computation in the JVM changed somehow between versions 7 and 8. But, if that happened, the JVM 8 change log does not seem to mention it (I did not look at the code of JVM though).

**Comment by mulugetam** [comment]
@mikism yes, the `System.identityHashCode(object)` implementation has changed. By default JDK 7 uses hashCode option 0 which calls `os::random` while JDK 8 defaults to hashCode 5 which uses Marsaglia's xor-shift PRNG algorithm with a thread-specific state. For my data sets, the problem occurs irrespective of which option I chose. 

There could be a room for improvement in the hash functions used inside Kryo to generate indices so that they create a better distribution. But so far they work just fine (as good as some other well known hash functions that I have tried).

FYI:

To check the default hashCode:
`java -XX:+UnlockDiagnosticVMOptions -XX:+PrintFlagsFinal -version | grep hash
`

The source (line 557): [http://hg.openjdk.java.net/jdk8u/jdk8u60/hotspot/file/37240c1019fd/src/share/vm/runtime/synchronizer.cpp ](url) 

Thanks,
Mulugeta

**Comment by mmammo** [comment]
@magro attached you may find the [jmh](http://openjdk.java.net/projects/code-tools/jmh/) benchmark results. Overall I attempted different ways to fix the solution:

1. increase the stash size - though this fix worked for some of our genome data sets, it failed for others. Increasing it by a factor of 8 (current is 2) didn't help either.

1. increase the number of push iterations:

    _changed this_: 

    `pushIterations = Math.max(Math.min(capacity, 8), (int)Math.sqrt(capacity) / 8)`

   _to this_:

   `pushIterations = Math.max(Math.min(capacity, 8), (int)Math.sqrt(capacity))`

   This helped avoid early stash fill ups (for data sets in the order of 2^10 - 2^20) but still failed. See attached result after running our genome data. 

1. change the hash functions - the current hash functions are fairly fast and seem to work as good as other standard hash functions. So, no change here.

1. add a fourth hash function:

  So far this is the best solution - it 
[kryo_genome_data.xlsx](https://github.com/EsotericSoftware/kryo/files/142058/kryo_genome_data.xlsx)
[raw_jmh_data.xlsx](https://github.com/EsotericSoftware/kryo/files/142059/raw_jmh_data.xlsx)

guarantees less collisions and Kryo will NOT break. It worked for all of our genome data sets as well as test random data set. 

 Since adding a hash function introduces a latency, I chose to include it for data sets greater than equal to 2^15 (32,768) elements.  A [jmh](http://openjdk.java.net/projects/code-tools/jmh/) benchmark indicates that for data sets less than 2^15 the latency is very low. 

 The _worst possible_ latency measured is 16ms and 12ms for inserting and reading 2^25 elements, respectively. Which amounts to a latency of ~0.5 ns per a put or a get operation, worst case. Which is very good. 

I will do a pull request soon. Let me know if you have any questions.

Thanks,
Mulugeta.

**Comment by magro** [comment]
Again, great analysis, looking forward to the PR :-)
Am 23.02.2016 08:17 schrieb "mmammo" <notifications@github.com>:

**Comment by anujlal01** [comment]
did you create pull request for this @mulugetam @magro

**Comment by mulugetam** [comment]
@anujlal01 I will do a PR as soon as I'm complete an internal process here to get a code out.

**Comment by magro** [comment]
Resolved with #400
